plugins/source/scheduler_round_robin.go

```diff
@@ -0,0 +1,99 @@
+package source
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"golang.org/x/sync/semaphore"
+)
+
+func (p *Plugin) syncRoundRobin(ctx context.Context, spec specs.Source, client schema.ClientMeta, tables schema.Tables, resolvedResources chan<- *schema.Resource) {
+	tableConcurrency := max(spec.Concurrency/minResourceConcurrency, minTableConcurrency)
+	resourceConcurrency := tableConcurrency * minResourceConcurrency
+
+	p.tableSems = make([]*semaphore.Weighted, p.maxDepth)
+	for i := uint64(0); i < p.maxDepth; i++ {
+		p.tableSems[i] = semaphore.NewWeighted(int64(tableConcurrency))
+		// reduce table concurrency logarithmically for every depth level
+		tableConcurrency = max(tableConcurrency/2, minTableConcurrency)
+	}
+	p.resourceSem = semaphore.NewWeighted(int64(resourceConcurrency))
+
+	// we have this because plugins can return sometimes clients in a random way which will cause
+	// differences between this run and the next one.
+	preInitialisedClients := make([][]schema.ClientMeta, len(tables))
+	for i, table := range tables {
+		clients := []schema.ClientMeta{client}
+		if table.Multiplex != nil {
+			clients = table.Multiplex(client)
+		}
+		preInitialisedClients[i] = clients
+		// we do this here to avoid locks so we initial the metrics structure once in the main goroutines
+		// and then we can just read from it in the other goroutines concurrently given we are not writing to it.
+		p.metrics.initWithClients(table, clients)
+	}
+
+	// We start a goroutine that logs the metrics periodically.
+	// It needs its own waitgroup
+	var logWg sync.WaitGroup
+	logWg.Add(1)
+
+	logCtx, logCancel := context.WithCancel(ctx)
+	go p.periodicMetricLogger(logCtx, &logWg)
+
+	var wg sync.WaitGroup
+	type tableClient struct {
+		table  *schema.Table
+		client schema.ClientMeta
+	}
+
+	// interleave table-clients so that we get:
+	// table1-client1, table2-client1, table3-client1, table1-client2, table2-client2, table3-client2, ...
+	tableClients := make([]tableClient, 0)
+	c := 0
+	for {
+		addedNew := false
+		for i, table := range tables {
+			if c < len(preInitialisedClients[i]) {
+				tableClients = append(tableClients, tableClient{table: table, client: preInitialisedClients[i][c]})
+				addedNew = true
+			}
+		}
+		c++
+		if !addedNew {
+			break
```

Comment:
Im not sure this will work. what if one table has less clients then the others? then it wont go to add others.

Reply:
The code was written to take that case into account, so I think it should work. The algorithm is roughly like this:

```
set counter to 1
add first client for all tables with >= 1 clients
set counter to 2
add second client for all tables with >= 2 clients
...
set counter to N
add Nth client for all tables with >= N clients
set counter to N+1
stop because no tables are found to have >= N+1 clients
```

I believe it handles all cases, but I've extracted this into a function now and added a test for it, so we can be sure.